File: benchmarks/references/ground-truth/R/tc-034-sufficient-followup.R
#!/usr/bin/env Rscript
#' TC-034: Sufficient Follow-up Assessment
#'
#' Assesses whether subjects have sufficient safety follow-up per protocol
#' (e.g., 90 days post-last-dose). Also computes median follow-up duration
#' using the reverse Kaplan-Meier method.
#'
#' Output:
#'   - N (%) with adequate follow-up by arm (>= threshold days post-last-dose)
#'   - Median follow-up duration (reverse KM) by arm with CI
#'   - N still ongoing, N discontinued, N died
#'   - Per-subject follow-up details
#'
#' Usage:
#'   Rscript tc-034-sufficient-followup.R --seed 42 --n 200 --output tc-034-output.json
#'   Rscript tc-034-sufficient-followup.R --data shared_adsl.csv --output tc-034-output.json --ars-output tc-034-ars.json

suppressPackageStartupMessages({
  library(jsonlite)
  library(survival)
})

# --- Argument parsing ---
args <- commandArgs(trailingOnly = TRUE)
seed <- 42
n_subjects <- 200
output_file <- ""
data_csv <- ""
ars_output <- ""
followup_threshold <- 90  # days post-last-dose

i <- 1
while (i <= length(args)) {
  if (args[i] == "--seed") { seed <- as.integer(args[i + 1]); i <- i + 2 }
  else if (args[i] == "--n") { n_subjects <- as.integer(args[i + 1]); i <- i + 2 }
  else if (args[i] == "--output") { output_file <- args[i + 1]; i <- i + 2 }
  else if (args[i] == "--data") { data_csv <- args[i + 1]; i <- i + 2 }
  else if (args[i] == "--ars-output") { ars_output <- args[i + 1]; i <- i + 2 }
  else if (args[i] == "--threshold") { followup_threshold <- as.integer(args[i + 1]); i <- i + 2 }
  else { i <- i + 1 }
}

cat(sprintf("[%s] TC-034 Sufficient Follow-up Assessment | seed=%d n=%d threshold=%dd\n",
    Sys.time(), seed, n_subjects, followup_threshold))

# --- Data generation ---
# ADSL-style: per-subject with randomization date, last dose date, last follow-up date, status
generate_followup_data <- function(n, seed) {
  set.seed(seed)
  n_exp <- n %/% 2
  n_ctl <- n - n_exp
  trt <- rep(c(1, 0), times = c(n_exp, n_ctl))
  trt <- sample(trt)

  # Randomization date (study start: 2023-01-01 + random offset)
  rand_date <- as.Date("2023-01-01") + sample(0:365, n, replace = TRUE)

  # Treatment duration (weeks): Exp ~ 24-60, Ctl ~ 12-36
  treat_dur_weeks <- ifelse(trt == 1,
    round(rnorm(n, mean = 42, sd = 10)),
    round(rnorm(n, mean = 24, sd = 7)))
  treat_dur_weeks <- pmax(treat_dur_weeks, 2)

  # Last dose date = randomization + treatment duration
  last_dose_date <- rand_date + treat_dur_weeks * 7

  # Follow-up duration post-last-dose (days)
  # Exp: ~60-180 days, Ctl: ~60-150 days (some will be < 90 = inadequate)
  fu_post_dose <- ifelse(trt == 1,
    round(rnorm(n, mean = 120, sd = 40)),
    round(rnorm(n, mean = 100, sd = 35)))
  fu_post_dose <- pmax(fu_post_dose, 0)

  # Last follow-up date
  last_fu_date <- last_dose_date + fu_post_dose

  # Subject status at end of study:
  # 1 = Ongoing (still in follow-up), 2 = Completed, 3 = Discontinued, 4 = Died
  status_prob <- matrix(c(
    0.20, 0.55, 0.15, 0.10,  # Exp
    0.15, 0.65, 0.12, 0.08   # Ctl
  ), nrow = 2, byrow = TRUE)

  status <- integer(n)
  for (j in 1:n) {
    status[j] <- sample(1:4, 1, prob = status_prob[trt[j] + 1, ])
  }

  # For died subjects, cap follow-up at a reasonable time
  died_idx <- which(status == 4)
  if (length(died_idx) > 0) {
    fu_post_dose[died_idx] <- pmin(fu_post_dose[died_idx],
      round(rnorm(length(died_idx), mean = 45, sd = 20)))
    fu_post_dose[died_idx] <- pmax(fu_post_dose[died_idx], 0)
    last_fu_date[died_idx] <- last_dose_date[died_idx] + fu_post_dose[died_idx]
  }

  # Follow-up from randomization (total)
  fu_from_rand <- as.numeric(last_fu_date - rand_date)

  # Adequate follow-up: >= threshold days post-last-dose AND not died
  adequate <- as.integer(fu_post_dose >= followup_threshold & status != 4)

  data.frame(
    USUBJID = sprintf("SUB%04d", seq_len(n)),
    TRT01PN = trt,
    TRT01A = ifelse(trt == 1, "Experimental", "Control"),
    RANDDT = rand_date,
    LASTDOSEDT = last_dose_date,
    LASTFUDT = last_fu_date,
    TREATDUR_W = treat_dur_weeks,
    FU_POST_DOSE = fu_post_dose,
    FU_FROM_RAND = fu_from_rand,
    ADEQUATE_FU = adequate,
    STATUS = status,
    STATUS_LABEL = c("ONGOING", "COMPLETED", "DISCONTINUED", "DIED")[status],
    stringsAsFactors = FALSE
  )
}

if (data_csv != "" && file.exists(data_csv)) {
  df <- read.csv(data_csv, stringsAsFactors = FALSE)
  # Convert date columns
  df$RANDDT <- as.Date(df$RANDDT)
  df$LASTDOSEDT <- as.Date(df$LASTDOSEDT)
  df$LASTFUDT <- as.Date(df$LASTFUDT)
  cat(sprintf("Loaded data: %s (%d rows)\n", data_csv, nrow(df)))
} else {
  df <- generate_followup_data(n_subjects, seed)
  cat(sprintf("Generated data: %d subjects\n", nrow(df)))
}

# --- Summary functions ---
summarize_cont <- function(x) {
  list(
    n = length(x),
    mean = round(mean(x), 4),
    sd = round(sd(x), 4),
    median = round(median(x), 4),
    min = round(min(x), 4),
    max = round(max(x), 4),
    q1 = round(quantile(x, 0.25), 4),
    q3 = round(quantile(x, 0.75), 4)
  )
}

# --- Reverse Kaplan-Meier for median follow-up ---
# Reverse KM: censor event deaths, treat everything else as events
# Using Fu_from_rand as time, event = 1 if not died (reverse), 0 if died
reverse_km_median <- function(time, status_label) {
  # In reverse KM, "event" = still alive (not died), "censored" = died
  event <- as.integer(status_label != "DIED")
  if (sum(event) == 0) return(list(median = NA_real_, ci_lower = NA_real_, ci_upper = NA_real_))
  surv_obj <- Surv(time, event)
  fit <- survfit(surv_obj ~ 1)
  med <- summary(fit)$table[["median"]]
  ci_lower <- summary(fit)$table[["0.95LCL"]]
  ci_upper <- summary(fit)$table[["0.95UCL"]]
  list(median = round(as.numeric(med), 4), ci_lower = round(as.numeric(ci_lower), 4), ci_upper = round(as.numeric(ci_upper), 4))
}

# --- Compute summaries ---
exp_df <- df[df$TRT01PN == 1, ]
ctl_df <- df[df$TRT01PN == 0, ]

# Adequate follow-up counts
adeq_exp <- sum(exp_df$ADEQUATE_FU)
adeq_ctl <- sum(ctl_df$ADEQUATE_FU)

# Status distribution
status_counts <- function(sub_df) {
  labels <- c("ONGOING", "COMPLETED", "DISCONTINUED", "DIED")
  sapply(labels, function(l) sum(sub_df$STATUS_LABEL == l))
}

status_exp <- status_counts(exp_df)
status_ctl <- status_counts(ctl_df)

# Reverse KM median follow-up
revkm_exp <- reverse_km_median(exp_df$FU_FROM_RAND, exp_df$STATUS_LABEL)
revkm_ctl <- reverse_km_median(ctl_df$FU_FROM_RAND, ctl_df$STATUS_LABEL)

# Follow-up post-dose summary
fu_post_exp <- summarize_cont(exp_df$FU_POST_DOSE)
fu_post_ctl <- summarize_cont(ctl_df$FU_POST_DOSE)

# Follow-up from randomization summary
fu_rand_exp <- summarize_cont(exp_df$FU_FROM_RAND)
fu_rand_ctl <- summarize_cont(ctl_df$FU_FROM_RAND)

result <- list(
  tc_id = "TC-034",
  tc_name = "Sufficient Follow-up Assessment",
  metadata = list(
    n_total = as.integer(nrow(df)),
    n_experimental = as.integer(nrow(exp_df)),
    n_control = as.integer(nrow(ctl_df)),
    population = "SAFETY",
    followup_threshold_days = as.integer(followup_threshold),
    date_unit = "days",
    r_version = R.version.string
  ),
  adequate_followup = list(
    experimental = list(
      n = as.integer(adeq_exp),
      pct = round(adeq_exp / nrow(exp_df) * 100, 2)
    ),
    control = list(
      n = as.integer(adeq_ctl),
      pct = round(adeq_ctl / nrow(ctl_df) * 100, 2)
    )
  ),
  status_distribution = list(
    experimental = list(
      ongoing = as.integer(status_exp["ONGOING"]),
      completed = as.integer(status_exp["COMPLETED"]),
      discontinued = as.integer(status_exp["DISCONTINUED"]),
      died = as.integer(status_exp["DIED"])
    ),
    control = list(
      ongoing = as.integer(status_ctl["ONGOING"]),
      completed = as.integer(status_ctl["COMPLETED"]),
      discontinued = as.integer(status_ctl["DISCONTINUED"]),
      died = as.integer(status_ctl["DIED"])
    )
  ),
  reverse_km_followup = list(
    experimental = revkm_exp,
    control = revkm_ctl
  ),
  fu_post_dose = list(
    experimental = fu_post_exp,
    control = fu_post_ctl
  ),
  fu_from_randomization = list(
    experimental = fu_rand_exp,
    control = fu_rand_ctl
  ),
  per_subject = lapply(seq_len(nrow(df)), function(i) {
    list(
      USUBJID = df$USUBJID[i],
      TRT01A = df$TRT01A[i],
      RANDDT = as.character(df$RANDDT[i]),
      LASTDOSEDT = as.character(df$LASTDOSEDT[i]),
      LASTFUDT = as.character(df$LASTFUDT[i]),
      TREATDUR_W = df$TREATDUR_W[i],
      FU_POST_DOSE = df$FU_POST_DOSE[i],
      FU_FROM_RAND = df$FU_FROM_RAND[i],
      ADEQUATE_FU = df$ADEQUATE_FU[i],
      STATUS = df$STATUS[i],
      STATUS_LABEL = df$STATUS_LABEL[i]
    )
  })
)

json_out <- toJSON(result, auto_unbox = TRUE, pretty = TRUE)

if (output_file != "") {
  writeLines(json_out, output_file)
  cat(sprintf("Written: %s\n", output_file))
} else {
  cat(json_out, "\n")
}

# --- ARS output ---
if (ars_output != "") {
  ars <- list(
    analysisResult = list(
      id = "TC-034",
      version = "1.0",
      analysisReason = "Assess sufficiency of safety follow-up per protocol requirements",
      analysisMethod = list(
        name = "Sufficient Follow-up Assessment",
        codeTemplate = "adequate = (fu_post_dose >= threshold) & (status != DIED)",
        parameters = list(
          list(name = "followup_threshold", role = "parameter", value = followup_threshold),
          list(name = "fu_post_dose", role = "input", source = "ADSL.LASTFUDT - ADSL.LASTDOSEDT"),
          list(name = "status", role = "input", source = "ADSL.EOSSTT")
        )
      ),
      analysisVariables = list(
        list(name = "TRT01A", role = "treatment", dataset = "ADSL"),
        list(name = "FU_POST_DOSE", role = "result", dataset = "derived"),
        list(name = "ADEQUATE_FU", role = "flag", dataset = "derived"),
        list(name = "STATUS_LABEL", role = "status", dataset = "ADSL")
      ),
      analysisPopulation = list(
        id = "SAFETY",
        filter = "SAFFL == 'Y'",
        n = as.integer(nrow(df))
      ),
      resultGroups = list(
        list(id = "EXP", label = "Experimental", n = as.integer(nrow(exp_df))),
        list(id = "CTL", label = "Control", n = as.integer(nrow(ctl_df)))
      ),
      analysisResultsData = list(
        list(groupId = "EXP", metric = "pct_adequate_fu", value = round(adeq_exp / nrow(exp_df) * 100, 2)),
        list(groupId = "CTL", metric = "pct_adequate_fu", value = round(adeq_ctl / nrow(ctl_df) * 100, 2)),
        list(groupId = "EXP", metric = "median_followup", value = revkm_exp$median),
        list(groupId = "CTL", metric = "median_followup", value = revkm_ctl$median)
      )
    )
  )
  writeLines(toJSON(ars, auto_unbox = TRUE, pretty = TRUE), ars_output)
  cat(sprintf("ARS written: %s\n", ars_output))
}

File: benchmarks/references/ground-truth/Python/tc_034_sufficient_followup.py
#!/usr/bin/env python3
"""TC-034: Sufficient Follow-up Assessment — Python

Assesses whether subjects have sufficient safety follow-up per protocol
(e.g., 90 days post-last-dose). Also computes median follow-up duration
using the reverse Kaplan-Meier method.

Output:
  - N (%) with adequate follow-up by arm (>= threshold days post-last-dose)
  - Median follow-up duration (reverse KM) by arm with CI
  - N still ongoing, N discontinued, N died
  - Per-subject follow-up details

Dependencies: numpy, pandas, lifelines
Usage: python tc_034_sufficient_followup.py --seed 42 --n 200 --output results.json
"""

import argparse
import json
import sys
from pathlib import Path
from datetime import date, timedelta

import numpy as np
import pandas as pd


def generate_followup_data(n: int, seed: int, threshold: int = 90) -> pd.DataFrame:
    """Generate ADSL-style follow-up dataset."""
    rng = np.random.default_rng(seed)
    n_exp = n // 2
    n_ctl = n - n_exp
    trt = np.array([1] * n_exp + [0] * n_ctl)
    rng.shuffle(trt)

    # Randomization date (study start: 2023-01-01 + random offset)
    rand_offset = rng.integers(0, 366, size=n)
    rand_date = np.array([date(2023, 1, 1) + timedelta(days=int(d)) for d in rand_offset])

    # Treatment duration (weeks)
    treat_dur_weeks = np.where(
        trt == 1,
        rng.normal(42, 10, size=n),
        rng.normal(24, 7, size=n),
    ).round().astype(int)
    treat_dur_weeks = np.maximum(treat_dur_weeks, 2)

    # Last dose date
    last_dose_date = np.array(
        [rand_date[i] + timedelta(weeks=int(treat_dur_weeks[i])) for i in range(n)]
    )

    # Follow-up post-last-dose (days)
    fu_post_dose = np.where(
        trt == 1,
        rng.normal(120, 40, size=n),
        rng.normal(100, 35, size=n),
    ).round().astype(int)
    fu_post_dose = np.maximum(fu_post_dose, 0)

    # Last follow-up date
    last_fu_date = np.array(
        [last_dose_date[i] + timedelta(days=int(fu_post_dose[i])) for i in range(n)]
    )

    # Subject status: 1=Ongoing, 2=Completed, 3=Discontinued, 4=Died
    status_prob = np.where(
        trt == 1,
        [0.20, 0.55, 0.15, 0.10],
        [0.15, 0.65, 0.12, 0.08],
    )
    status = np.array([
        rng.choice(4, p=status_prob[trt[i]]) + 1 for i in range(n)
    ])

    # For died subjects, cap follow-up
    died_mask = status == 4
    if died_mask.any():
        fu_post_dose[died_mask] = np.maximum(
            rng.normal(45, 20, size=died_mask.sum()).round().astype(int),
            0,
        )
        last_fu_date = np.array(
            [last_dose_date[i] + timedelta(days=int(fu_post_dose[i])) for i in range(n)]
        )

    # Follow-up from randomization (total days)
    fu_from_rand = np.array(
        [(last_fu_date[i] - rand_date[i]).days for i in range(n)]
    )

    # Adequate follow-up: >= threshold days post-last-dose AND not died
    adequate = ((fu_post_dose >= threshold) & (status != 4)).astype(int)

    status_labels = np.where(status == 1, "ONGOING",
                     np.where(status == 2, "COMPLETED",
                     np.where(status == 3, "DISCONTINUED", "DIED")))

    return pd.DataFrame({
        "USUBJID": [f"SUB{i:04d}" for i in range(1, n + 1)],
        "TRT01PN": trt.astype(int),
        "TRT01A": np.where(trt == 1, "Experimental", "Control"),
        "RANDDT": rand_date,
        "LASTDOSEDT": last_dose_date,
        "LASTFUDT": last_fu_date,
        "TREATDUR_W": treat_dur_weeks,
        "FU_POST_DOSE": fu_post_dose,
        "FU_FROM_RAND": fu_from_rand,
        "ADEQUATE_FU": adequate,
        "STATUS": status,
        "STATUS_LABEL": status_labels,
    })


def summarize_cont(x):
    arr = np.asarray(x, dtype=float)
    return {
        "n": int(len(arr)),
        "mean": round(float(np.mean(arr)), 4),
        "sd": round(float(np.std(arr, ddof=1)), 4),
        "median": round(float(np.median(arr)), 4),
        "min": round(float(np.min(arr)), 4),
        "max": round(float(np.max(arr)), 4),
        "q1": round(float(np.percentile(arr, 25)), 4),
        "q3": round(float(np.percentile(arr, 75)), 4),
    }


def reverse_km_median(time, status_label):
    """Reverse KM: event = alive (not DIED), censored = died.

    Returns median survival time and CI (Brookmeyer-Crowley method):
    - Lower CI: first time where upper CI of S(t) drops below 0.5
    - Upper CI: last time where lower CI of S(t) stays above 0.5
    """
    from lifelines import KaplanMeierFitter

    event = (status_label != "DIED").astype(int)
    if event.sum() == 0:
        return {"median": None, "ci_lower": None, "ci_upper": None}

    kmf = KaplanMeierFitter()
    kmf.fit(time, event_observed=event)

    med = kmf.median_survival_time_
    if np.isinf(med) or np.isnan(med):
        return {"median": None, "ci_lower": None, "ci_upper": None}

    med = float(med)

    # Brookmeyer-Crowley CI for median:
    # Find times where the CI bands of S(t) cross 0.5
    ci_df = kmf.confidence_interval_survival_function_
    # lifelines gives: index = timeline, columns = lower/upper bounds of S(t)
    # Column names: like 'KM_lower_0.95' and 'KM_upper_0.95'
    col_lower = ci_df.columns[0]  # lower CI of S(t)
    col_upper = ci_df.columns[1]  # upper CI of S(t)

    # Brookmeyer-Crowley CI for median:
    # Lower CI of median: last time where lower CI of S(t) >= 0.5
    # (the latest time where the lower bound still hasn't crossed 0.5)
    lower_cross = ci_df.index[ci_df[col_lower] >= 0.5]
    ci_lower = float(lower_cross[-1]) if len(lower_cross) > 0 else None

    # Upper CI of median: first time where upper CI of S(t) <= 0.5
    # (the earliest time where even the upper bound has crossed 0.5)
    upper_cross = ci_df.index[ci_df[col_upper] <= 0.5]
    ci_upper = float(upper_cross[0]) if len(upper_cross) > 0 else None

    return {
        "median": round(med, 4),
        "ci_lower": round(ci_lower, 4) if ci_lower is not None else None,
        "ci_upper": round(ci_upper, 4) if ci_upper is not None else None,
    }


def main():
    parser = argparse.ArgumentParser(description="TC-034: Sufficient Follow-up Assessment")
    parser.add_argument("--seed", type=int, default=42)
    parser.add_argument("--n", type=int, default=200)
    parser.add_argument("--output", type=str, default="")
    parser.add_argument("--data", type=str, default="")
    parser.add_argument("--data-csv", type=str, default="", dest="data_csv")
    parser.add_argument("--ars-output", type=str, default="")
    parser.add_argument("--threshold", type=int, default=90)
    args = parser.parse_args()

    data_path = args.data or args.data_csv
    threshold = args.threshold

    print(f"[{pd.Timestamp.now()}] TC-034 Sufficient Follow-up Assessment | seed={args.seed} n={args.n} threshold={threshold}d")

    if data_path and Path(data_path).exists():
        df = pd.read_csv(data_path)
        df["RANDDT"] = pd.to_datetime(df["RANDDT"]).dt.date
        df["LASTDOSEDT"] = pd.to_datetime(df["LASTDOSEDT"]).dt.date
        df["LASTFUDT"] = pd.to_datetime(df["LASTFUDT"]).dt.date
        print(f"Loaded data: {data_path} ({len(df)} rows)")
    else:
        df = generate_followup_data(args.n, args.seed, threshold)
        print(f"Generated data: {len(df)} subjects")

    exp_df = df[df["TRT01PN"] == 1]
    ctl_df = df[df["TRT01PN"] == 0]

    # Adequate follow-up counts
    adeq_exp = int(exp_df["ADEQUATE_FU"].sum())
    adeq_ctl = int(ctl_df["ADEQUATE_FU"].sum())

    # Status distribution
    def status_counts(sub_df):
        return {
            "ongoing": int((sub_df["STATUS_LABEL"] == "ONGOING").sum()),
            "completed": int((sub_df["STATUS_LABEL"] == "COMPLETED").sum()),
            "discontinued": int((sub_df["STATUS_LABEL"] == "DISCONTINUED").sum()),
            "died": int((sub_df["STATUS_LABEL"] == "DIED").sum()),
        }

    status_exp = status_counts(exp_df)
    status_ctl = status_counts(ctl_df)

    # Reverse KM median follow-up
    revkm_exp = reverse_km_median(exp_df["FU_FROM_RAND"].values, exp_df["STATUS_LABEL"].values)
    revkm_ctl = reverse_km_median(ctl_df["FU_FROM_RAND"].values, ctl_df["STATUS_LABEL"].values)

    # Follow-up post-dose summary
    fu_post_exp = summarize_cont(exp_df["FU_POST_DOSE"])
    fu_post_ctl = summarize_cont(ctl_df["FU_POST_DOSE"])

    # Follow-up from randomization summary
    fu_rand_exp = summarize_cont(exp_df["FU_FROM_RAND"])
    fu_rand_ctl = summarize_cont(ctl_df["FU_FROM_RAND"])

    result = {
        "tc_id": "TC-034",
        "tc_name": "Sufficient Follow-up Assessment",
        "metadata": {
            "n_total": int(len(df)),
            "n_experimental": int(len(exp_df)),
            "n_control": int(len(ctl_df)),
            "population": "SAFETY",
            "followup_threshold_days": int(threshold),
            "date_unit": "days",
            "python_version": sys.version,
        },
        "adequate_followup": {
            "experimental": {
                "n": adeq_exp,
                "pct": round(adeq_exp / len(exp_df) * 100, 2),
            },
            "control": {
                "n": adeq_ctl,
                "pct": round(adeq_ctl / len(ctl_df) * 100, 2),
            },
        },
        "status_distribution": {
            "experimental": status_exp,
            "control": status_ctl,
        },
        "reverse_km_followup": {
            "experimental": revkm_exp,
            "control": revkm_ctl,
        },
        "fu_post_dose": {
            "experimental": fu_post_exp,
            "control": fu_post_ctl,
        },
        "fu_from_randomization": {
            "experimental": fu_rand_exp,
            "control": fu_rand_ctl,
        },
        "per_subject": [
            {
                "USUBJID": row["USUBJID"],
                "TRT01A": row["TRT01A"],
                "RANDDT": str(row["RANDDT"]),
                "LASTDOSEDT": str(row["LASTDOSEDT"]),
                "LASTFUDT": str(row["LASTFUDT"]),
                "TREATDUR_W": int(row["TREATDUR_W"]),
                "FU_POST_DOSE": int(row["FU_POST_DOSE"]),
                "FU_FROM_RAND": int(row["FU_FROM_RAND"]),
                "ADEQUATE_FU": int(row["ADEQUATE_FU"]),
                "STATUS": int(row["STATUS"]),
                "STATUS_LABEL": row["STATUS_LABEL"],
            }
            for _, row in df.iterrows()
        ],
    }

    json_out = json.dumps(result, indent=2, default=str)

    if args.output:
        Path(args.output).write_text(json_out)
        print(f"Written: {args.output}")
    else:
        print(json_out)

    # ARS output
    if args.ars_output:
        ars = {
            "analysisResult": {
                "id": "TC-034",
                "version": "1.0",
                "analysisReason": "Assess sufficiency of safety follow-up per protocol requirements",
                "analysisMethod": {
                    "name": "Sufficient Follow-up Assessment",
                    "codeTemplate": "adequate = (fu_post_dose >= threshold) & (status != DIED)",
                    "parameters": [
                        {"name": "followup_threshold", "role": "parameter", "value": threshold},
                        {"name": "fu_post_dose", "role": "input", "source": "ADSL.LASTFUDT - ADSL.LASTDOSEDT"},
                        {"name": "status", "role": "input", "source": "ADSL.EOSSTT"},
                    ],
                },
                "analysisVariables": [
                    {"name": "TRT01A", "role": "treatment", "dataset": "ADSL"},
                    {"name": "FU_POST_DOSE", "role": "result", "dataset": "derived"},
                    {"name": "ADEQUATE_FU", "role": "flag", "dataset": "derived"},
                    {"name": "STATUS_LABEL", "role": "status", "dataset": "ADSL"},
                ],
                "analysisPopulation": {
                    "id": "SAFETY",
                    "filter": "SAFFL == 'Y'",
                    "n": int(len(df)),
                },
                "resultGroups": [
                    {"id": "EXP", "label": "Experimental", "n": int(len(exp_df))},
                    {"id": "CTL", "label": "Control", "n": int(len(ctl_df))},
                ],
                "analysisResultsData": [
                    {"groupId": "EXP", "metric": "pct_adequate_fu", "value": round(adeq_exp / len(exp_df) * 100, 2)},
                    {"groupId": "CTL", "metric": "pct_adequate_fu", "value": round(adeq_ctl / len(ctl_df) * 100, 2)},
                    {"groupId": "EXP", "metric": "median_followup", "value": revkm_exp["median"]},
                    {"groupId": "CTL", "metric": "median_followup", "value": revkm_ctl["median"]},
                ],
            }
        }
        Path(args.ars_output).write_text(json.dumps(ars, indent=2, default=str))
        print(f"ARS written: {args.ars_output}")


if __name__ == "__main__":
    main()
